### PR TITLE
FEATURE: Add a return code to node:repair command

### PR DESCRIPTION
**Pull request by zaveryukha**
This adds a return code to the `node:repair` command in node command controller plugin functions.

### CODE REVIEW

**Comment by kdambekalns** [comment]
**How I did it**
Added to all checks return true in case of error.

**How to verify it**
run command - ./flow node:repair --dry-run=true
if errors exist you'll see output like this:
During run, some error were found!!!

**Checklist**

- [X] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)

File: Neos.ContentRepository/Classes/Command/NodeCommandControllerPlugin.php [hunk 1]
```diff
@@ -223,10 +223,11 @@ public static function getSubCommandDescription($controllerCommandName)
      * @param boolean $cleanup If false, cleanup tasks are skipped
      * @param string $skip Skip the given check or checks (comma separated)
      * @param string $only Only execute the given check or checks (comma separated)
-     * @return void
+     * @return boolean
      */
     public function invokeSubCommand($controllerCommandName, ConsoleOutput $output, NodeType $nodeType = null, $workspaceName = 'live', $dryRun = false, $cleanup = true, $skip = null, $only = null)
```
Comment (daniellienert) [inline Neos.ContentRepository/Classes/Command/NodeCommandControllerPlugin.php]:
```suggestion
    public function invokeSubCommand($controllerCommandName, ConsoleOutput $output, NodeType $nodeType = null, $workspaceName = 'live', $dryRun = false, $cleanup = true, $skip = null, $only = null): bool
```

File: Neos.ContentRepository/Classes/Command/NodeCommandController.php [hunk 2]
```diff
@@ -78,7 +78,7 @@ class NodeCommandController extends CommandController implements DescriptionAwar
      * @param boolean $cleanup If false, cleanup tasks are skipped
      * @param string $skip Skip the given check or checks (comma separated)
      * @param string $only Only execute the given check or checks (comma separated)
-     * @return void
+     * @return boolean
```
Comment (daniellienert) [inline Neos.ContentRepository/Classes/Command/NodeCommandController.php]:
```suggestion
     * @return void
```

File: Neos.ContentRepository/Classes/Command/NodeCommandController.php [hunk 3]
```diff
@@ -106,16 +106,24 @@ public function repairCommand($nodeType = null, $workspace = 'live', $dryRun = f
...
-            $plugin->invokeSubCommand('repair', $this->output, $nodeType, $workspace, $dryRun, $cleanup, $skip, $only);
+            if ($plugin->invokeSubCommand('repair', $this->output, $nodeType, $workspace, $dryRun, $cleanup, $skip, $only) === true) {
+                $hasErrors = true;
+            }
             $this->outputLine();
         }
 
         $this->outputLine('Node repair finished.');
+        if ($hasErrors === true) {
+            $this->outputLine('During run, some error were found!');
```
Comment (daniellienert) [inline Neos.ContentRepository/Classes/Command/NodeCommandController.php]:
```suggestion
            $this->outputLine('During run, some errors were found!');
```

File: Neos.ContentRepository/Classes/Command/NodeCommandControllerPlugin.php [hunk 4]
```diff
@@ -223,7 +228,7 @@ public static function getSubCommandDescription($controllerCommandName)
      * @param boolean $cleanup If false, cleanup tasks are skipped
      * @param string $skip Skip the given check or checks (comma separated)
      * @param string $only Only execute the given check or checks (comma separated)
-     * @return void
+     * @return boolean
```
Comment (daniellienert) [inline Neos.ContentRepository/Classes/Command/NodeCommandControllerPlugin.php]:
```suggestion
     * @return void
```

File: Neos.ContentRepository/Classes/Command/NodeCommandControllerPlugin.php [hunk 5]
```diff
@@ -1312,6 +1333,10 @@ protected function repairShadowNodes($workspaceName, $dryRun, NodeType $nodeType
             $fixedShadowNodes !== 1 ? 's' : ''
         ]);
 
+        if (fixedShadowNodes>0) {
```
Comment (daniellienert) [inline Neos.ContentRepository/Classes/Command/NodeCommandControllerPlugin.php]:
```suggestion
        if (fixedShadowNodes > 0) {
```

**Comment by bwaidelich** [comment]
This makes a lot of sense but the implementation clashes a little with #2264 .
Instead of `invokeSubCommand()` returning `int` or `bool` it should IMO return a `Neos\Error\Messages\Result`. We could use that to determine errors (`$result->hasErrors()`) and to collect status information about the invoked command.
I'll try to prepare a 3rd PR to give it a try

**Comment by bwaidelich** [comment]
FYI: Using the `Result` didn't work out because it would break the interface. But I found a way with #2278

**Comment by jonnitto** [comment]
What is the status here?

**Comment by bwaidelich** [comment]
With #2278 `node:repair` sends a proper exit code (1) if errors occurred.
Errors include:
* Missing/unknown node types
* SQL errors when fixing identifiers
It does _not_ include expected conditions (like missing child nodes that need to be created).
For those the `dry-run` flag can be used.

All in all the mentioned PR is a foundation for further improvements. i.e. it would be easy to add `verbose` and/or `quiet` flag or some different (machine readable) output format.

I think this one can be closed. Do you agree, @zaveryukha ?

**Comment by kdambekalns** [comment]
Thanks @zaveryukha for starting this… even though it turned around like that, you triggered it to become better! ❤️

**Comment by zaveryukha** [comment]
Thanks @bwaidelich @kdambekalns for you support. Looking forward to enhance this feature in a proper way

**Pull request closed without merging**